test/proper_lib.erl

```diff
@@ -12,11 +12,23 @@
 codec(Mod, Opts, Data) ->
     codec(Mod, Opts, Data, fun(V) -> V end).
 
-codec(Mod, Opts, Data, Canonical) ->
+% Encode and then decode the Input
+encode_decode(Mod, Opts, Input) ->
     {_, Config} = Mod:init(Opts),
     TypeInfo = #type_info{config = Config},
-    <<Size:32, Encoded:Size/binary>> = iolist_to_binary(Mod:encode(Data, TypeInfo)),
-    Canonical(Data) =:= Canonical(Mod:decode(Encoded, TypeInfo)).
+    <<Size:32, Encoded:Size/binary>> = iolist_to_binary(Mod:encode(Input, TypeInfo)),
+    Mod:decode(Encoded, TypeInfo).
+
+% Passing a function will apply that function to the input Data and
+% encoded/decoded Data before comparison.
+codec(Mod, Opts, Data, Canonical) when is_function(Canonical) ->
+    Decoded = encode_decode(Mod, Opts, Data),
+    Canonical(Data) =:= Canonical(Decoded);
+% Passing anything else will simply compare the Output with the encoded/decoded
+% Input
+codec(Mod, Opts, Data, Output) ->
```

Comment:
This version is more convenient in the property test. Without this, I need a more complicated function to deal with the round-trip. Something like,

```
f({{_X, _Y, _Z}, {_A, _B, _C}, {_D, _E}} = Timestamp) -> apply_offset(Timestamp);
f({{_X, _Y, _Z}, {_A, _B, _C} = Timestamp} -> Timestamp.
```

It isn't that complicated, but it is much easier to just provide the expected output.